### PR TITLE
Perf test for color filter with saveLayer

### PR DESCRIPTION
**Pull request by liyuqian**
This should reveal the problem in:
https://github.com/flutter/flutter/issues/51975

Fixing that issue should significantly lower the frame rasterization
time.

### CODE REVIEW

**Comment by liyuqian** [comment]
@karhu: I've simplified your reproduction app, and put it inside Flutter's devicelab perf test suite. Do you think that you can directly create a PR like this next time?

As for the two problem you mentioned,
1. "saveLayer with ColorFilter disables the raster cache" will be tracked in https://github.com/flutter/flutter/issues/51975 and a fix would show a significant improvement in this test.
2. "picture is always recreated under fade transition" is actually something that you can fix in your code. You simply need to add a `RepaintBoundary` [here](1) to not let the fade animation rebuild everything below.

[1]: https://github.com/flutter/flutter/pull/52063/files#diff-6f5343f40209ca92bf82ca21c5cbaab4R64

File: dev/benchmarks/macrobenchmarks/lib/src/color_filter_and_fade.dart [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+  final Shadow shadow;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    final Rect rect = Offset.zero & size;
+
+    final Paint paint = Paint();
+    if (color != null) {
+      paint.colorFilter = ColorFilter.mode(shadow.color, BlendMode.srcIn);
```
Comment (flar) [inline dev/benchmarks/macrobenchmarks/lib/src/color_filter_and_fade.dart]: You are testing color for null, but then using shadow.color here. Does color get used as anything other than a flag?

File: dev/benchmarks/macrobenchmarks/lib/src/color_filter_and_fade.dart [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    canvas.saveLayer(null, paint);
+    canvas.translate(shadow.offset.dx, shadow.offset.dy);
+    canvas.drawRect(rect, Paint());
+    canvas.drawRect(rect, Paint()..maskFilter = MaskFilter.blur(BlurStyle.normal, shadow.blurSigma));
+    canvas.restore();
+
+    canvas.drawRect(rect, Paint()..color = color?? Colors.black);
```
Comment (flar) [inline dev/benchmarks/macrobenchmarks/lib/src/color_filter_and_fade.dart]: Ah, here color gets used...

**Comment by karhu** [comment]
@liyuqian Thanks a lot for your work on this! I went over your changes and it  is indeed a great reference for future repro apps. Will definitely give it a try next time :+1:

File: dev/benchmarks/macrobenchmarks/lib/src/color_filter_and_fade.dart (hunk 2 above)
Comment (liyuqian) [reply dev/benchmarks/macrobenchmarks/lib/src/color_filter_and_fade.dart]: So does this look good to you?

**Comment by flar** [comment]
The use of the color param as a flag to trigger the use of shadow.color looks a little wonky, but that's not a blocking concern.

LGTM.

**Review by flar — APPROVED** [review]

**Comment by liyuqian** [comment]
Oh, I see it. Now I've removed `color` and changed everything to `useColorFilter` now. Thanks for the suggestion!

**Pull request merged**